crates/typst/src/layout/grid/layout.rs

```diff
@@ -216,38 +226,123 @@ impl CellGrid {
+                    .at(span)?
+            };
+            let x = resolved_index % c;
+            let y = resolved_index / c;
+
+            // Let's resolve the cell so it can determine its own fields
+            // based on its final position.
+            let cell = cell.resolve_cell(
+                x,
+                y,
+                &fill.resolve(engine, x, y)?,
+                align.resolve(engine, x, y)?,
+                inset,
+                styles,
+            );
+
+            // Now let's check if the cell's position is valid.
+            if let Some(existing_cell) = resolved_cells.get_mut(resolved_index) {
+                // We are trying to place a cell in an existing position.
+                // Ensure we aren't trying to place a cell where there is
+                // already one.
+                if existing_cell.is_some() {
+                    bail!(
+                        span,
+                        "attempted to place two distinct cells at column {x}, row {y}";
+                        hint: "try specifying your cells in a different order"
+                    );
+                }
+
+                // Ok, position is available, so let's place the cell here.
+                *existing_cell = Some(cell);
+            } else if resolved_index == resolved_cells.len() {
+                // We can just place the new cell at the end of the grid vector.
+                // No other cell can be there.
+                resolved_cells.push(Some(cell));
+            } else {
+                // Here, resolved_index > resolved_cells.len(). Thus, the cell
+                // needs to be placed in a position which doesn't exist yet in
+                // the grid (out of bounds). We will add enough absent
+                // positions for this to be possible. They must be absent as no
+                // cells actually occupy them (they can be overridden later);
+                // however, if no cells occupy them as we finish building the
+                // grid, then such positions will be replaced by empty cells.
+                let new_position_count = resolved_index - resolved_cells.len();
+                resolved_cells.extend(
+                    std::iter::repeat_with(|| None)
+                        .take(new_position_count)
+                        .chain(std::iter::once(Some(cell))),
+                );
+            }
+        }
+
+        // If not all columns in the last row have cells, we will add absent
+        // positions (later converted to empty cells) and complete the row so
+        // that those positions are susceptible to show rules and receive grid
+        // styling.
+        // We apply '% c' twice so that 'cells_remaining' is zero when the last
+        // row is already filled (then 'resolved_cells.len() % c' would be
+        // zero).
+        let cells_remaining = (c - resolved_cells.len() % c) % c;
```

Comment:
We could technically even integrate this into the resize above to ensure that we always push full rows.

Reply:
I considered doing this, but I thought that it doesn't matter that much since we are about to allocate new space for a vector anyway, so if we allocate extra above, it would be a bit of a waste. I also considered that allocating extra space in the loop could be better in terms of performance (as in, we would expect that, if a cell is placed in a certain row, then we'd expect the row to be filled later), but:
1. We already allocate enough to fit the initial cell count at the start, so that'd be a no-op in most cases;
2. Rust is already good at allocating extra space when we don't know how much the vector will expand (which is precisely our case here when arbitrary positioning is used).

Either way, there's no way to confirm that without benchmarks :p. Additionally, one argument in favor of resizing in the loop is that we could assume the invariant that the vector always has full rows, and thus (in `resolve_cell_position`) not have to consider the case where the next available position is out of bounds, other than for cells which request a particular column (then the next available spot in the column could be in a row that doesn't exist in the vector yet).

So we should probably ponder about this a bit. Perhaps it's better for the PR as a whole if we do this change...

Reply:
Okay, well, I did it. In the end I was kinda neutral about it, and there were no performance implications apparently, so I guess it's ok to do it, no hard feelings though.